src/ledger/AccountFrame.cpp

```diff
@@ -118,10 +118,115 @@ AccountFrame::getBalance() const
     return (mAccountEntry.balance);
 }
 
+int64_t
+getBuyingLiabilities(AccountEntry const& acc, LedgerManager const& lm)
+{
+    assert(lm.getCurrentLedgerVersion() >= 10);
+    return (acc.ext.v() == 0) ? 0 : acc.ext.v1().liabilities.buying;
+}
+
+int64_t
+getSellingLiabilities(AccountEntry const& acc, LedgerManager const& lm)
+{
+    assert(lm.getCurrentLedgerVersion() >= 10);
+    return (acc.ext.v() == 0) ? 0 : acc.ext.v1().liabilities.selling;
+}
+
+int64_t
+AccountFrame::getBuyingLiabilities(LedgerManager const& lm) const
+{
+    return stellar::getBuyingLiabilities(mAccountEntry, lm);
+}
+
+int64_t
+AccountFrame::getSellingLiabilities(LedgerManager const& lm) const
+{
+    return stellar::getSellingLiabilities(mAccountEntry, lm);
+}
+
+bool
+AccountFrame::addBuyingLiabilities(int64_t delta, LedgerManager const& lm)
+{
+    assert(lm.getCurrentLedgerVersion() >= 10);
+    assert(getBalance() >= 0);
+    if (delta == 0)
+    {
+        return true;
+    }
+    int64_t buyingLiab = (mAccountEntry.ext.v() == 0) ? 0 : mAccountEntry.ext.v1().liabilities.buying;
+
+    int64_t maxLiabilities = INT64_MAX - getBalance();
+    bool res = stellar::addBalance(buyingLiab, delta, maxLiabilities);
+    if (res)
+    {
+        if (mAccountEntry.ext.v() == 0)
+        {
+            mAccountEntry.ext.v(1);
+            mAccountEntry.ext.v1().liabilities = Liabilities{0, 0};
+        }
+        mAccountEntry.ext.v1().liabilities.buying = buyingLiab;
+    }
+    return res;
+}
+
+bool
+AccountFrame::addSellingLiabilities(int64_t delta, LedgerManager const& lm)
+{
+    assert(lm.getCurrentLedgerVersion() >= 10);
+    assert(getBalance() >= 0);
+    if (delta == 0)
+    {
+        return true;
+    }
+    int64_t sellingLiab = (mAccountEntry.ext.v() == 0) ? 0 : mAccountEntry.ext.v1().liabilities.selling;
+
+    int64_t maxLiabilities = getBalance() - getMinimumBalance(lm);
+    if (maxLiabilities < 0)
+    {
+        return false;
+    }
+
+    bool res = stellar::addBalance(sellingLiab, delta, maxLiabilities);
+    if (res)
+    {
+        if (mAccountEntry.ext.v() == 0)
+        {
+            mAccountEntry.ext.v(1);
+            mAccountEntry.ext.v1().liabilities = Liabilities{0, 0};
+        }
+        mAccountEntry.ext.v1().liabilities.selling = sellingLiab;
+    }
+    return res;
+}
+
 bool
-AccountFrame::addBalance(int64_t delta)
+AccountFrame::addBalance(int64_t delta, LedgerManager const& lm)
 {
-    return stellar::addBalance(mAccountEntry.balance, delta);
+    if (delta == 0 && lm.getCurrentLedgerVersion() >= 10)
```

Comment:
why is currentLedgerVersion checked here?

Reply:
Good question, I see now that it doesn't need to be checked. Fixed.